### PR TITLE
Call MDTF from ADF Revised 20240612

### PR DESCRIPTION
**Pull request by bitterbark**
- Call [MDTF](https://mdtf-diagnostics.readthedocs.io/en/main/) from ADF

This adds a section in the config and run files so the ADF can directly call the [MDTF (Model Diagnostics Task Force (MDTF)-Diagnostics package) ](https://www.gfdl.noaa.gov/mdtf-diagnostics/).

This is the simplest possible implementation that calls an existing, installed version of the MDTF using existing, installed conda envs. For now, timeseries files will be copied to the requested MDTF directory/filenames. Imminent data catalogue functionality in MDTF might change this.

Basic documentation of new functionality:
in config.yaml, there is a new section:

> 
> diag_mdtf_info:
> mdtf_run: false [default]

If `mdtf_run:true:`
in run_adf_diag
>
> #Call the MDTF:
> if diag.get_mdtf_info('mdtf_run'):
> diag.setup_run_mdtf()

[Detailed documentation of implementation/call tree]
(https://www2.cgd.ucar.edu/cms/bundy/Projects/diagnostics/mdtf/notes/run_mdtf_from_adf.html#Documentation)

- Allow hist_str to be a list, and to be different for each cam case (and baseline)
The config.yaml file variable can now be a list, and is specified within the `diag_cam_climo` and `diag_cam_baseline_climo` lists
>
>hist_str: [cam.h2,cam.h0]

These timeseries files are all placed in the same ts/ directory, which may need to be modified for multiple component hist strings. (The MDTF checks file types for time frequency for its own purposes).

### CODE REVIEW

**Comment by bitterbark** [comment]
I would like to request a pass on the linter errors. I've fixed all the ones that came from my code. In between last time I merged and tried for the PR and this time, it seems that code has been committed successfully without resolving any of these issues. They are not in my code, and the work I put in last time to clean up the code just made it even more difficult to merge in the branch this time.  Can we please get this PR in and then do the cleanup separately so we can all start from a clean slate?

**Comment by nusbaume** [comment]
@bitterbark I have gone ahead and fixed the linter errors/warnings.  I'll try to give this PR one last look over, and if OK on my end will send it over to @justin-richling to run a test or two.  Thanks!

**Comment by bitterbark** [comment]
Thanks, Jesse (and Justin). I’m on vacation until June 25 so I can work then on anything that comes up in the testing.On Jun 13, 2024, at 13:47, Jesse Nusbaumer ***@***.***> wrote:﻿
@bitterbark I have gone ahead and fixed the linter errors/warnings.  I'll try to give this PR one last look over, and if OK on my end will send it over to @justin-richling to run a test or two.  Thanks!

File: config_amwg_default_plots.yaml [hunk 1]
```diff
@@ -290,6 +296,71 @@ diag_cvdp_info:
...
+    # Run the MDTF on the model cases
+    mdtf_run: false
+
+    # The file that will be written by ADF to input to MDTF. Call this whatever you want.
+    mdtf_input_settings_filename : mdtf_input.json   
+
+    ## MDTF code path, sets the location of the MDTF codebase and pre-compiled conda envs
+    #  CHANGE if you have any: your own MDTF code, installed conda envs and/or obs_data
+
+    mdtf_codebase_path : /glade/campaign/cgd/amp/amwg/mdtf
```
Comment (nusbaume) [inline config_amwg_default_plots.yaml:329]: Just FYI I think we will want to make this path and the files below it globally readable (like `ADF_obs`).  Currently I think it is only readable by the `cgd` group.

**Review by nusbaume — APPROVED** [review]
This PR now looks good to me!  @bitterbark There was one issue about the MDTF directory permissions on Derecho that sadly I am not allowed to fix, but it can easily wait until whenever you are back at work again.  Thanks for all your effort!

File: lib/adf_info.py [hunk 2]
```diff
@@ -211,15 +228,15 @@ def __init__(self, config_file, debug=False):
...
                     eyear_baseline = found_eyear_baseline
-            #End if
-
-            #Check if history file path exists:
-            if baseline_hist_locs:
+            # End if
 
+            # Check if history file path exists:
+            if any(baseline_hist_locs):
+                hist_str = baseline_hist_str[0]
```
Comment (justin-richling) [inline lib/adf_info.py:235]: This fails if the user only supplies a string, should it be run through the same function as the test case(s) using `hist_str_to_list` or do it inline 

```suggestion
                if not isinstance(conf_val, list):
                    baseline_hist_str = [baseline_hist_str]
                hist_str = baseline_hist_str[0]
```

or above where `baseline_hist_str` is declared?

**Review by justin-richling — CHANGES_REQUESTED** [review]
@bitterbark This is great, thanks for the work to integrate the MDTF. I just tested it for an ADF run only and I only had one issue about the baseline hist locs, see my comment in `adf_info.py`. Other than that it seems like the ADF works as intended with this addition.

File: lib/adf_info.py (hunk 2 above)
Comment (bitterbark) [reply lib/adf_info.py:243]: Justin, I'm having trouble getting the changes you sent me to merge. I'm assuming this suggestion was included before you did the wider merge?

**Comment by bitterbark** [comment]
I took the merged code from Justin but am running into errors when I try to push it to my branch. 
First question: it looks like the changes Justin made might already be committed to [my branch](https://github.com/bitterbark/ADF/tree/call_mdtf_from_adf.20240612). Is there a chance I don't need to make any other mods other than the one suggestion?  Unfortunately this thread shows extensive merge conflicts that i don't see how to resolve.

So when I try to push the new files from Justin up to my branch, I get merge errors. Honestly not different from what happened earlier this month, when it said my online branch was ahead of the local branch (somehow) and then I let it do the merge and it messed everything up and I had to redo all my changes by hand. 

Wondering if one of you has advice. 

> 
> git push
>  ! [rejected]        call_mdtf_from_adf.20240612 ->
>  call_mdtf_from_adf.20240612 (fetch first)
> error: failed to push some refs to 'github.com:bitterbark/ADF.git'
> hint: Updates were rejected because the remote contains work that you do
> hint: not have locally. This is usually caused by another repository
>  pushing
> hint: to the same ref. You may want to first integrate the remote changes
> hint: (e.g., 'git pull ...') before pushing again.
> hint: See the 'Note about fast-forwards' in 'git push --help' for details.

So then I try 

> git fetch
> ...
> Unpacking objects: 100% (47/47), 37.01 KiB | 2.00 KiB/s, done.
> From github.com:bitterbark/ADF
>    63c2c90..3c541f2  call_mdtf_from_adf.20240612 ->
>   origin/call_mdtf_from_adf.20240612
> 

and 

>  git pull
> hint: You have divergent branches and need to specify how to reconcile  them.
> hint: You can do so by running one of the following commands sometime before
> hint: your next pull:
> hint:
> hint:   git config pull.rebase false  # merge
> hint:   git config pull.rebase true   # rebase
> hint:   git config pull.ff only       # fast-forward only
> 

I really don't want to have to merge by hand again. Surely there is some way git can do this. Advice?

**Comment by nusbaume** [comment]
@bitterbark @justin-richling I have resolved the merge conflicts and gotten the linting tests to pass again.  The merge was a little dicey so please retest this branch with your relevant ADF workflow whenever you get the chance.  Assuming everything works for you then I think (?) this PR is good to go.  Thanks!

**Comment by bitterbark** [comment]
Thanks so much Jesse!

**Pull request merged**